airbyte-integrations/bases/standard-destination-test/src/main/java/io/airbyte/integrations/standardtest/destination/comparator/AdvancedTestDataComparator.java

```diff
@@ -0,0 +1,191 @@
+    LOGGER.info("Expected data {}", expected);
+    LOGGER.info("Actual data   {}", actual);
+    assertEquals(expected.size(), actual.size());
+    final Iterator<JsonNode> expectedIterator = expected.iterator();
+    final Iterator<JsonNode> actualIterator = actual.iterator();
+    while (expectedIterator.hasNext() && actualIterator.hasNext()) {
+      compareObjects(expectedIterator.next(), actualIterator.next());
+    }
+  }
+
+  protected void compareObjects(final JsonNode expectedObject, final JsonNode actualObject) {
+    LOGGER.info("Expected Object : {}", expectedObject);
+    LOGGER.info("Actual Object   : {}", actualObject);
+    if (areBothEmpty(expectedObject, actualObject)) {
+      LOGGER.info("Both rows are empty.");
+    } else {
+      final Iterator<Map.Entry<String, JsonNode>> expectedDataIterator = expectedObject.fields();
+      assertEquals(expectedObject.size(), actualObject.size(), "Unequal row size");
+      while (expectedDataIterator.hasNext()) {
+        final Map.Entry<String, JsonNode> expectedEntry = expectedDataIterator.next();
+        final JsonNode expectedValue = expectedEntry.getValue();
+        JsonNode actualValue = null;
+        String key = expectedEntry.getKey();
+        for (final String tmpKey : resolveIdentifier(expectedEntry.getKey())) {
+          actualValue = actualObject.get(tmpKey);
+          if (actualValue != null) {
+            key = tmpKey;
+            break;
+          }
+        }
+        LOGGER.info("For {} Expected {} vs Actual {}", key, expectedValue, actualValue);
+        assertTrue(actualObject.has(key));
+        assertSameValue(expectedValue, actualValue);
+      }
+    }
+  }
+
+  private boolean areBothEmpty(final JsonNode expectedData, final JsonNode actualData) {
+    if (expectedData.size() == 0 && actualData.size() == 0) {
+      return true;
+    } else if (expectedData.size() <= 1 && actualData.size() <= 1) {
+      boolean isExpectedEmpty = (expectedData.size() == 0 || expectedData.iterator().next().asText().isEmpty());
+      return isExpectedEmpty && (actualData.size() == 0 || actualData.iterator().next().asText().isEmpty());
+    } else {
+      return false;
+    }
+  }
+
+  // Allows subclasses to implement custom comparison asserts
+  protected void assertSameValue(final JsonNode expectedValue, final JsonNode actualValue) {
+    LOGGER.info("assertSameValue : {} vs {}", expectedValue, actualValue);
+
+    assertTrue(compareJsonNodes(expectedValue, actualValue), "Expected value " + expectedValue + " vs Actual value " + actualValue);
+  }
+
+  protected boolean compareJsonNodes(final JsonNode expectedValue, final JsonNode actualValue) {
+    if (expectedValue == null || actualValue == null)
+      return expectedValue == null && actualValue == null;
+    else if (expectedValue.isNumber() || expectedValue.isDouble() || expectedValue.isFloat())
+      return compareNumericValues(expectedValue.asText(), actualValue.asText());
```

Comment:
why do these values need to be converted to text?

Reply:
The idea is to compare all numbers in the one common way. The method `compareNumericValues` should parse all values into one format and compare them.
Otherwise we should take into account different cases like:

- methods which returns primitive types can return default value. additional check for empty element.
- expected and actual jsons returns different types (int vs double)